### PR TITLE
EntityManager and Storage connection

### PR DESCRIPTION
**Pull request by iK4tsu**
Changes:
 * fixed methods not being `pure`
 * refactored `add` method to `set`
 * refactored the constraint of a component to only accept mutable fields
 * added the Assoc Array of StorageInfo[TypeInfo] searchable by the Component type to have access to the Storage through the EntityManager
 * added the following methods to manipulate entities and components:
   * `get` - fetches a component from an entity, returns a pointer to the component or null if it fails
   * `set` - associates an entity with a component, returns true if successful or false if it fails
   * `remove` - removes a component from an entity, returns true if successful or false if it fails
   * `removeAll` - removes all components from an entity or storage, returns true if successful or false if it fails
   * `gen` - now an entity can be generated with components on the fly, returns the newly generated entity
   * `getOrSet` - set the component if the entity is no associated with one and/or returns it
   * `size` - gets the the amount of components/entities within a component storage
 
A component can be associated to an entity the following ways:
```d
@Component struct Foo {}
@Component struct Bar {}
auto em = new EntityManager!size_t();

// with set
em.set!Foo(em.gen());
em.set(Foo, Bar)(em.gen());
em.set(em.gen(), Foo());
em.set(em.gen(), Foo(), Bar());

// with gen
em.gen!Foo;
em.gen!(Foo, Bar);
em.gen(Foo());
em.gen(Foo(), Bar());

// with getOrSet
em.getOrSet!Foo(em.gen());
em.getOrSet!Foo(em.gen(), Foo());

// gen was also updated to work with EntityBuilder
em.entityBuilder
    .gen!(Foo)
    .gen!(Foo, Bar)
    .gen(Foo())
    .gen(Foo(), Bar());
```

Components can be fetched the following ways:
```d
@Component struct Foo { int x, y; }
auto em = new EntityManager!size_t();

// with get
Foo* foo = em.get!Foo(em.gen!Foo);

// with getOrSet
Foo* foo = em.getOrSet!Foo(e);
Foo* foo = em.getOrSet(e, Foo(2, 3));
```

Components can be removed using:
```d
@Component struct Foo {}
@Component struct Bar {}
@Component struct Foobar{}
auto em = new EntityManager!size_t();

em.remove!Foo(em.gen!Foo); // removes Foo, returns true
em.remove!Bar(em.gen!Foo); // does not remove any, returns false
em.removeAll(em.gen!(Foo, Bar)); // removes all from the newly gen entity, returns true
em.removeAll(em.entityNull); // invalid entity, returns false
em.removeAll!Foo(); // removes all components and entities in Foo Storage, returns true
em.removeAll!Foobar(); // there is no Foobar Storage, returns false
```

### CODE REVIEW

**Comment by iK4tsu** [comment]
The only aspect I'm not realy sure about is the `remove` logic when it comes to removing more than one component at the same time.

File: source/ecs/entity.d [hunk 1]
```diff
@@ -231,6 +236,71 @@ public:
...
+	 * Examples:
+	 * --------------------
+	 * @Component struct Foo { int x; } // x gets default initialized to int.init
+	 * auto em = new EntityManager!size_t;
+	 * em.gen!Foo(); // generates a new entity with Foo.init values
+	 * --------------------
+	 *
+	 * Params: component = a valid component.
+	 *
+	 * Returns: a newly generated Entity!T.
```
Comment (ljmf00) [inline source/ecs/entity.d:252]: It's fine, but you don't need to seperate single lines on documentation

**Review by ljmf00 — CHANGES_REQUESTED** [review]

File: source/ecs/entity.d (hunk 1 above)
Comment (iK4tsu) [reply source/ecs/entity.d:252]: It's just my style. I find it easier to read the written documentation this way.

**Comment by ljmf00** [comment]
> (...) if no Component is passed, like the current behavior, all components are removed?

I guess that shouldn't happen. It could lead to silent errors and not readable for outside people. `removeAll` is good enough to remove all entities. It's clear and not exaggeratedly long.

**Comment by iK4tsu** [comment]
> > (...) if no Component is passed, like the current behavior, all components are removed?
> 
> I guess that shouldn't happen. It could lead to silent errors and not readable for outside people. `removeAll` is good enough to remove all entities. It's clear and not exaggeratedly long.

`remove` does not remove entities. It removes components associated to an entity. The tricky thing here is:
 * `remove!(A)(e)` - removes `A` if associated with `e`,  returns **true** if successful
 * `remove!(A,B,...)(e)` - removes **only** the associated components **ignoring** the non associated, **only** returns **false** if the entity is invalid meaning **as long as the entity is valid** it **always** returns **true** even if no component passed was removed

**Pull request merged**